File: examples/sample/store.rs
pub struct UserStore {
    pub name: String,
}

impl UserStore {
    pub fn new(name: impl Into<String>) -> Self {
        Self { name: name.into() }
    }

    pub fn find_by_id(&self, id: &str) -> String {
        let user = self.load_user(id);
        self.render(user)
    }

    fn load_user(&self, id: &str) -> String {
        format!("{}#{}", self.name, id)
    }

    fn render(&self, user: String) -> String {
        format!("[{}]", user)
    }
}

pub fn render_user(store: &UserStore, id: &str) -> String {
    store.find_by_id(id)
}

File: examples/sample/store.c
#include <stdio.h>
#include <string.h>

typedef struct {
    int id;
    char name[64];
} User;

static User load_user(int id) {
    User u;
    u.id = id;
    strncpy(u.name, "Ada", sizeof(u.name));
    return u;
}

User find_by_id(int id) {
    return load_user(id);
}

void render_user(int id) {
    User u = find_by_id(id);
    printf("%d: %s\n", u.id, u.name);
}
